File: src/linkedin_action_center/ingestion/fetchers.py
"""Fetchers for LinkedIn Ads entity data: accounts, campaigns, creatives.

Each function accepts a ``LinkedInClient`` and returns raw dicts straight from
the API.  No transformation or persistence happens here.
"""

from __future__ import annotations

from linkedin_action_center.ingestion.client import LinkedInClient


def fetch_ad_accounts(client: LinkedInClient) -> list[dict]:
    """
    Return all ad accounts the authenticated member has access to.

    Endpoint: GET /rest/adAccounts?q=search
    """
    accounts = client.get_all_pages("/adAccounts", "q=search")
    return accounts


def fetch_campaigns(
    client: LinkedInClient,
    account_id: int,
    statuses: list[str] | None = None,
) -> list[dict]:
    """
    Return all campaigns for *account_id*.

    By default fetches ACTIVE, PAUSED, and DRAFT campaigns (excludes
    ARCHIVED and CANCELED to keep the dataset relevant).

    Endpoint: GET /rest/adAccounts/{id}/adCampaigns?q=search
    """
    if statuses is None:
        statuses = ["ACTIVE", "PAUSED", "DRAFT"]

    # LinkedIn REST API uses Restli query syntax: search=(status:(values:List(A,B,C)))
    status_list = ",".join(statuses)
    params = f"q=search&search=(status:(values:List({status_list})))"
    return client.get_all_pages(f"/adAccounts/{account_id}/adCampaigns", params)


def fetch_creatives(
    client: LinkedInClient,
    account_id: int,
    campaign_ids: list[int] | None = None,
) -> list[dict]:
    """
    Return all creatives for *account_id*, optionally filtered to specific campaigns.

    Endpoint: GET /rest/adAccounts/{id}/creatives?q=criteria
    """
    params = "q=criteria&sortOrder=ASCENDING"

    if campaign_ids:
        urns = ",".join(
            f"urn%3Ali%3AsponsoredCampaign%3A{cid}" for cid in campaign_ids
        )
        params += f"&campaigns=List({urns})"

    creatives = client.get_all_pages(
        f"/adAccounts/{account_id}/creatives", params,
    )
    return creatives

File: node-app/src/ingestion/fetchers.ts
/**
 * Fetchers for LinkedIn Ads entity data: accounts, campaigns, creatives.
 *
 * Each function accepts a LinkedInClient and returns raw dicts straight from
 * the API. No transformation or persistence happens here.
 */

import { LinkedInClient } from "./client.js";

export async function fetchAdAccounts(
  client: LinkedInClient,
): Promise<Record<string, unknown>[]> {
  return client.getAllPages("/adAccounts", "q=search");
}

export async function fetchCampaigns(
  client: LinkedInClient,
  accountId: number,
  statuses: string[] = ["ACTIVE", "PAUSED", "DRAFT"],
): Promise<Record<string, unknown>[]> {
  // LinkedIn REST API uses Restli query syntax
  const statusList = statuses.join(",");
  const params = `q=search&search=(status:(values:List(${statusList})))`;
  return client.getAllPages(`/adAccounts/${accountId}/adCampaigns`, params);
}

export async function fetchCreatives(
  client: LinkedInClient,
  accountId: number,
  campaignIds?: number[],
): Promise<Record<string, unknown>[]> {
  let params = "q=criteria&sortOrder=ASCENDING";

  if (campaignIds && campaignIds.length > 0) {
    const urns = campaignIds
      .map((cid) => `urn%3Ali%3AsponsoredCampaign%3A${cid}`)
      .join(",");
    params += `&campaigns=List(${urns})`;
  }

  return client.getAllPages(`/adAccounts/${accountId}/creatives`, params);
}
